questions.md

```diff
@@ -2,29 +2,72 @@
+
+```bash
+/usr/bin/time -v
+```  
+
 3. In order to be a "good citizen", you need to have some idea of much RAM your job requires.  In particular, you need
    to know the "peak" (i.e., maximum) RAM required at any point during execution.  Show an example of the shell command
    that you would use on a Linux machine to measure run time and peak ram usage of an arbitrary command, where the time/peak RAM values are written to a file.
-4. What are the units of your answer for number 3?
+answer:
+The shell command I would use would be:
+```bash
+/usr/bin/time -o stats.txt my_program
+```
+
+4. What are the units of your answer for number 3? 
+answer: kbytes
 5. What are the bash commands for the following operations:
 
-    * Checking that a file exists
-    * Checking that a file exists and is not empty
+    * Checking that a file exists: 
+answer:
+```bash
+[ -f  test.txt ] && echo "yes" || echo "no"
+```
+
+   * Checking if a file exists and is not empty
+answer:
+```bash
+[ -s test.txt ] && echo "exists and not empty" || echo "empty or doesn't exist"
+```
 
 6. How would you use the commands from your answer to 5 to write a work flow for HPC that only runs a job if the
    expected output file is **not** present.
+```bash
+ [ -f test.txt ] || sh test_script.sh
+```
 
 ## Trickier questions
 
 7. Would your answer to number 3 work on Apple's OS X operating system?  If no, do you have any idea why not? 
 
+answer: no, /usr/bin/time is linux specific. There is another program called time on OS X that is a different program and gives a different output.
+
 8. Most of the HPC nodes have 512Gb (gigabytes) of RAM. Let's say you have a job that will require **no more** than 24Gb
    of RAM.  How would you request resources so that you can run more than one job on a node **and** not cause nodes to
    crash?  Show an example of a skeleton HPC script as part of your answer.  **Knowing how to do this is super important
    and will save you loads of frustration and prevent you from taking out your colleagues' jobs on the cluster,
    preventing you from getting nasty emails from Harry!!!!!!!!!!!**
+
+```bash
+#!/bin/bash
+
+#$ -N log
+#$ -q bio
+#$ -l node=name_of_node
+#$ -l h_vmem=24GB
+
+echo "Write your commands here"
+```
+
+after submitting this script you can submit another shell script using #$ -l node=name_of_node by specifying the same node using #$ -l h_vmem=RAM# as long as the h_vmem does not add up to over the amount of RAM the node has.
```

Comment:
By default and as a common courtesy it is best to take the total amount of RAM on a system, subtract a small overhead (our admins have mostly taken care of this so we don't really need to worry about this value much) and then divide it by the total number of cores. if your job will be using 256GB of RAM on a 512GB RAM node it is best to also use the same portion of cores for your job, even though your job will only run on 1 core. Technically, our grid should have the option to limit the amount of maximum memory requested based on number of cores requested. i.e on a 64 core 512GB node the user should allocate a maximum of 8GB per core. if the single core job were to use 16GB then it would require that 2 cores be allocated thereby avoiding issues caused by the over consumption of RAM by a job.